Added Blog Post

michaelstaib opened the pull request.

Urigo commented on website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md:
```diff
@@ -0,0 +1,592 @@
+
+## Lets share and compete
+
+Late last year ChilliCream and The Guild met in Paris and discussed their approaches twoards distributed GraphQL. It became very clear that both companies were solving similar problems and we decided to join forces on this project. ChilliCream would provide the initial work on the Fusion spec and implementation while The Guild would start specifying their work on Mash and Open API and help shape  ...
```
```suggestion
Late last year ChilliCream and [The Guild](https://the-guild.dev/) met in Paris and discussed their approaches twoards distributed GraphQL. It became very clear that both companies were solving similar problems and we decided to join forces on this project. ChilliCream would provide the initial work on the Fusion spec and implementation while The Guild would start specifying their work on the [GraphQL Mesh Gateway](https://the-guild.dev/graphql/mesh) with [OpenAPI support](https://the-guild.dev/graphql/mesh/docs/handlers/openapi) and [Hive](https://the-guild.dev/graphql/hive) as an open schema registry to help shape the initial Fusion spec.
```

YassinEldeeb replied:
```suggestion
Late last year ChilliCream and The Guild met in Paris and discussed their approaches towards distributed GraphQL. It became very clear that both companies were solving similar problems and we decided to join forces on this project. ChilliCream would provide the initial work on the Fusion spec and implementation while The Guild would start specifying their work on Mash and Open API and help shape the initial Fusion spec.
```

Urigo commented on website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md:
```diff
@@ -0,0 +1,592 @@
+
+Late last year ChilliCream and The Guild met in Paris and discussed their approaches twoards distributed GraphQL. It became very clear that both companies were solving similar problems and we decided to join forces on this project. ChilliCream would provide the initial work on the Fusion spec and implementation while The Guild would start specifying their work on Mash and Open API and help shape  ...
+
+Today, we are thrilled to unveil GraphQL-Fusion an open specification under the MIT license. This initiative empowers everyone to craft tools and solutions centered around distributed GraphQL services. Complementing this announcement, we're also introducing Hot Chocolate Fusion, an early implementation of the GraphQL-Fusion spec draft.
```
```suggestion
Today, we are thrilled to unveil GraphQL-Fusion an open specification under the **MIT license**. This initiative empowers everyone to craft tools and solutions centered around distributed GraphQL services. Complementing this announcement, we're also introducing Hot Chocolate Fusion, an early implementation of the GraphQL-Fusion spec draft.
```

Urigo replied:
Should you also mention here the working group?

Urigo replied:
and are you ok with adding something like "And The Guild commitment to integrate Fusion into their existing gateway solutions" or something like that?

Urigo commented on website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md:
```diff
@@ -0,0 +1,592 @@
+
+## Conclusion
+
+We are still working on GraphQL-Fusion, but you can try it today with the Hot Chocolate Fusion Gateway. We plan to release the GraphQL-Fusion spec as it matures later this year under the MIT license. While the current iterations are primarily concerned with GraphQL, The Guild has begun specifying the Open API to GraphQL transformation they've developed in Mesh, and we'll integrate this as it beco ...
```
I would give the part where you can take any Apollo Subgraph and seamlessly migrate it to Fusion its own headline and paragraph
Something around seamless migration

michaelstaib replied:
I will ... also want to add an example snippet of that.

Urigo reviewed (changes requested):
this is soooooo exciting!!

michaelstaib commented on website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md:
```diff
@@ -0,0 +1,592 @@
+
+## Going Further
+
+While the subgraph inference of the schema composition is quite powerfull there are a lot of cases where we can further by declaring semntics of a schema. We do not need to integrate such annotations into our subgraph schema directly but can pass an additional GraphQL documents into the schema composition that hold type extensions with additional directive annotations.
```
```suggestion
While the subgraph inference of the schema composition is quite powerful there are a lot of cases where we can go further by declaring semantics of a schema. We do not need to integrate such annotations into our subgraph schema directly but can pass an additional GraphQL documents into the schema composition that hold type extensions with additional directive annotations.
```

n1ru4l commented on website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md:
```diff
@@ -0,0 +1,592 @@
+
+## Reshaping things
+
+When we rethink a bit our shipping subgraph we actually should realize that the `deliveryEstimate` does not reall need to be on a product type as the argument has clear requirements which are expressed by it's field arguments. Instead of having the field `deliveryEstimate` on the `Product` type itself it could very well be exposed through the `Query` type. At least in the context of our subgraph.
```
```suggestion
When we rethink a bit our shipping subgraph we actually should realize that the `deliveryEstimate` does not really need to be on a product type as the argument has clear requirements which are expressed by its field arguments. Instead of having the field `deliveryEstimate` on the `Product` type itself, it could be exposed through the `Query` type. At least in the context of our subgraph.
```

n1ru4l commented on website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md:
````diff
@@ -0,0 +1,592 @@
+}
+```
+
+In the case our subgraph is isolated and does not fully integrate with our though for model that we want to expose on the gateway we can also reshape it.
````
```suggestion
In the case, our subgraph is isolated and does not fully integrate with our though for the model that we want to expose on the gateway we can also reshape it.
```

n1ru4l commented on website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md:
```diff
@@ -0,0 +1,592 @@
+
+In the case our subgraph is isolated and does not fully integrate with our though for model that we want to expose on the gateway we can also reshape it.
+
+All the annotations can be put into separate graphql document providing the extending metadata. First lets make the whole query type private, we actually do not want to include anything by default form this subgraph.
```
```suggestion
All the annotations can be put into separate graphql documents providing the extending metadata. First lets make the whole query type private, we actually do not want to include anything by default form this subgraph.
```

n1ru4l commented on website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md:
```diff
@@ -0,0 +1,603 @@
+
+At its heart, GraphQL Fusion pivots around two foundational principles: schema composition and query planning.
+
+But before delving into these concepts, it's essential to retrace our steps. Let's revisit the challenges surfaced when people first tried GraphQL as a Gateway for constructing their GraphQL servers over REST APIs.
```
```suggestion
But before delving into these concepts, it's essential to retrace our steps. Let's revisit the challenges that surfaced when people first tried GraphQL as a Gateway for constructing their GraphQL servers over REST APIs.
```

michaelstaib replied:
I think the that is not needed in this sentence.

n1ru4l commented on website/src/blog/2023-08-15-fusion/2023-08-15-fusion.md:
```diff
@@ -0,0 +1,603 @@
+
+## Apollo Federation
+
+We recognize that some of you may have opted into Apollo Federation as a solution, and that is why we designed the Fusion schema composition so that it can compose any Apollo Federation subgraph into a Fusion subgraph. This allows you to seamlessly migration from an Apollo Federation setup to a GraphQL-Fusion setup without the need to rewrite everything.
```
```suggestion
We recognize that some of you may have opted into Apollo Federation as a solution, and that is why we designed the Fusion schema composition so that it can compose any Apollo Federation subgraph into a Fusion subgraph. This allows you to seamlessly migrate from an Apollo Federation setup to a GraphQL-Fusion setup without the need to rewrite everything.
```

michaelstaib commented:
@rstaib we need a banner for this blog post.

huv1k commented:
I just wonder, what about the subscription? There is nothing mentioning it 🙈

michaelstaib commented:
@huv1k I added a section for this with the latest commit ... Fusion supports all GraphQL features from defer and stream to subscription over to facebook style batching with export directives or multipart requests for file upload ... I added now a section to mention this. 

There are so many more features in fusion ... I fear if we put them all into this initial post it becomes to complex to comprehend for the reader.

michaelstaib commented:
This is the new section:

The Fusion query plan is another standardized component that tooling (like [Banana Cake Pop](https://eat.bananacakepop.com)) can use to give you insights into how efficiently the gateway can resolve the requested data.

![Banana Cake Pop - Query Plan Viewer](bcp-1.png)
*Also available in black ;)*

The Fusion Query plan consists of the following query plan node kinds: `Compose`, `If`, `Introspect`, `Parallel`, `Resolve`, `ResolveByKeyBatch`, `ResolveNode`, `Sequence`, and `Subscribe`. With these abstract nodes, the query planner is able to create complex execution plans that support every GraphQL feature and best practice right out of the gate.

While the `Fetch` and `Batch` nodes are clear about what they do in our plan, the compose step might be a mystery to you. In essence, the query planner can fetch data that does not align with the current structure of the request. Compose will take in the raw data fetched by resolve nodes and composes it into the GraphQL request structure. It also ensures that result coercion rules are correctly applied to be GraphQL spec-compliant.

![Banana Cake Pop - Query Plan Viewer](bcp-5.png)
*In this case, compose creates the result of a single selection set from multiple resolve nodes.*

The Hot Chocolate Fusion Gateway implementation supports for subscriptions already all supported subscription protocols, from the legacy Apollo subscription protocol over graphql-ws to graphql-sse.

Further, it supports file uploads with the GraphQL multipart request protocol, Facebook-style batching with the `@export` directives, and many more.

Distributed GraphQL should **not limit** what you can do with GraphQL.

michaelstaib commented:
Important to understand here is that the fusion spec defines how to do subscription but the gateways implementing this are free to support whichever protocol they see fit. In the case of Hot Chocolate we added the three main protocols including graphql-sse.
